Warnings fixes

ws233 opened the pull request:
Xcode 7 shows some extra warnings, which was hidden in Xcode 6. So this patch fixes them.
1. "G8Tesseract.mm:56:17: Method override for the designated initializer of the superclass '-init' not found" warning has been fixed. The default initializer implementation has been added.
2. Init is no longer unavailable. Instead it initializes the totally default Tesseract instance.
3. Unit test for the default initializer has been added.
4. "InitializationTests.m:265:49: Initialization of pointer of type 'id' to null from a constant boolean expression" warning has been fixed.

kevincon commented:
Can you investigate why coveralls reported that our coverage decreased from 100% to 64.557%?

ws233 commented:
@kevincon, I don't know the reason. It seems it was just a bug in coveralls. It showed that all the files were covered 100%, but total statistic was 64.557%.
I've just restarted the job and now there is 100% totally.
Pls, review. Thx!

kevincon commented:
I'm all for fixing warnings, but if the `init` function with no parameters is used to create a `G8Tesseract` object and then recognition is performed without setting a language (i.e. it's `nil`), then Tesseract defaults to looking for the English trained data file. That's what happened in this issue that prompted me to prevent the use of `init` without setting a language: #145 

How about a compromise? Can you make a log message be printed out if the `init` function with no parameters is used to say something like:

> WARNING: Initializing G8Tesseract without a language defaults to English, so make sure you either set the language afterward or have eng.traineddata in your tessdata folder, otherwise Tesseract will crash!

ws233 commented:
1. @kevincon, I've missed this ticket. Sorry.
I think log message is a goof idea. But I'll put it in the language setter. Since such situation may happen after the user has been changed the language manually as well.

2. Also there is an ambiguity in Tesseract initialization code and language setter. From one side we return nil from the initializer if we cannot init engine (so there is no tess instance at all). From other side if we try to change language from an already initialized tesseract instance and fail to initialize with it we'll have a tesseract wrapper with bad engine (that's an error actually and even if you try to change the language to a correct one later, tesseract will crash since engine is corrupted).

So I've changed the behavior, so it's similar in both cases above. An initializer now always returns a tesseract wrapper even if it cannot initialize engine. But in the fail case it will set language to nil and no engine alive will exist. In the case the users try to set wrong language using the setter they will also get a nil languages property and no tesseract engine alive. So in both cases the users are able to check the reason and recover with correct language.

@kevincon, pls, refer to the PR.
and let me know if I missed something. Thx!

ws233 commented:
I've also found that all the functions, which need tesseract engine to work properly never checks that the engine is valid. So they may crash in case the engine doesn't exist. So I've added those conditions to implementation and notes to the header. I've also will add unit tests to cover new code soon.

kevincon commented:
Okay, let me know when you add the unit tests so I can review the PR.

ws233 commented:
According to the whole idea of returning valid instance of the wrapper always, but instead have a dead engine if error occurs, I've changed the code, which returns nil, if the user tries to copy tessdata files from app bundle to the absoluteDataPath.
So we may make absoluteDataPath writable, so the user is able to change tessdata path on an existing tesseract wrapper instance :)

@kevincon, actually, I'm done with making everything similar in this PR. May also make absoluteDataPath writable if you wish, but I don't think it's so useful. Pls, review very carefully, since I've changed very basic functionality significantly. Thx!

ws233 commented:
@kevincon, is there something wrong with the PR? Can you review it and merge, if everything is ok?

kevincon commented:
Sorry for the delay, things have been really busy at work. I'll start reviewing the PR now.

kevincon commented on TesseractOCR/G8Tesseract.h:
```diff
@@ -260,18 +275,26 @@ extern NSInteger const kG8MaxCredibleResolution;
-- (instancetype)init __attribute__((unavailable("Use initWithLanguage:language instead")));
+/**
+ *  The default initializer.
+ *  @return A G8Tesseract instance initialized with niether language,
```
Typo: "niether" -> neither

kevincon commented on TesseractOCR/G8Tesseract.h:
```diff
@@ -260,18 +275,26 @@ extern NSInteger const kG8MaxCredibleResolution;
+ *  @return The initialized Tesseract object. 
+ *
+ *  @note   You should check that the language property match the languages 
+ *          you have specified. Also you may ensure that tesseract is properly
```
It looks like the convention we've been using is to capitalize "Tesseract", so can you change any instances of "tesseract" in these files to "Tesseract"?

kevincon commented on TesseractOCR/G8Tesseract.h:
```diff
@@ -260,18 +275,26 @@ extern NSInteger const kG8MaxCredibleResolution;
+ *
+ *  @note   You should check that the language property match the languages 
+ *          you have specified. Also you may ensure that tesseract is properly
+ *          initialized by checking isEngineConfigured property.
```
Also can you wrap any properties and function names like "isEngineConfigured" in the comments with backtick marks so they'll be formatted as code in the documentation?

kevincon commented on TesseractOCR/G8Tesseract.h:
```diff
@@ -279,10 +302,14 @@ extern NSInteger const kG8MaxCredibleResolution;
- *  @return The initialized Tesseract object, or `nil` if there was an error.
+ *  @return The initialized Tesseract object. 
+ *
+ *  @note   You should check that the language property match the languages
```
"match" -> matches

kevincon commented on TesseractOCR/G8Tesseract.h:
```diff
@@ -307,14 +334,18 @@ extern NSInteger const kG8MaxCredibleResolution;
- *  @return The initialized Tesseract object, or `nil` if there was an error.
+ *  @return The initialized Tesseract object.
+ *
+ *  @note   You should check that the language property match the languages
```
"match" -> matches

kevincon commented on TesseractOCR/G8Tesseract.h:
```diff
@@ -347,13 +378,17 @@ extern NSInteger const kG8MaxCredibleResolution;
- *  @return The initialized Tesseract object, or `nil` if there was an error.
+ *  @return The initialized Tesseract object.
+ *
+ *  @note   You should check that the language property match the languages
```
"match" -> matches

kevincon commented on TesseractOCR/G8Tesseract.mm:
```diff
@@ -287,18 +307,28 @@ - (void)setVariableValue:(NSString *)value forKey:(NSString *)key
+    
     [self resetFlags];
 
+    if (!value) {
```
Since you check `value`, shouldn't we also check `key`? The function could just return early if `key` isn't a valid string so that it's a no-op.

ws233 replied:
I thought about it, but I chose the native iOS behavior, which is crashing if key is nil.
Just try `setValue:forKey:` with a nil key to any object. 
Also in xCode 7 function parameters are now can be nullable and nonnull. So if we later specify none keyword for the key parameter the static analyzer will not allow the user to send nil there. I'll do this later in the xCode 7 branch.

kevincon replied:
Okay sounds good.

kevincon commented on TesseractOCR/G8Tesseract.mm:
```diff
@@ -310,18 +340,124 @@ - (void)setVariablesFromDictionary:(NSDictionary *)dictionary
+- (void)setEngineImage:(UIImage *)image {
+    
+    if (image.size.width <= 0 || image.size.height <= 0) {
+        NSLog(@"ERROR: Image has not size!");
```
I know you just moved this code, but can you change this log message to be "Image has invalid size!"?

kevincon commented on TesseractOCR/G8Tesseract.mm:
```diff
@@ -310,18 +340,124 @@ - (void)setVariablesFromDictionary:(NSDictionary *)dictionary
+        pixDestroy(&pix);
+    }
+    
+    _image = image;
```
I don't think we should save the image if the engine isn't configured or if we're not successful in calling `_tesseract->SetImage(pix)`. Can you rework this function so that's not the case?

ws233 replied:
I've done this to follow the whole idea, that all the Tesseract parameters once set, shouldn't be reset after any engine reinitialization. I mean, if I set variables, rect, image, resolution, blacklist or whitelist, I shouldn't reset them again after changing the language or recognition mode. That's how all other properties works now, so I suppose that the `image` property should work the same. That allows all our properties to be independent from the engine. What do you think?

kevincon replied:
Ah I see, yeah I guess that's okay, thanks for explaining it to me.

kevincon commented on TesseractOCR/G8Tesseract.mm:
```diff
@@ -310,18 +340,124 @@ - (void)setVariablesFromDictionary:(NSDictionary *)dictionary
+        
+        _language = language.copy;
+        if (!self.language) {
+            NSLog(@"WARNING: Setting G8Tesseract language to nill defaults to English, so make sure you either set the language afterward or have eng.traineddata in your tessdata folder, otherwise Tesseract will crash!");
```
nill -> nil

kevincon commented on TesseractOCR/G8Tesseract.mm:
```diff
@@ -310,18 +340,124 @@ - (void)setVariablesFromDictionary:(NSDictionary *)dictionary
-    if ([_language isEqualToString:language] == NO) {
-        _language = [language copy];
-
+    if ([language isEqualToString:_language] == NO || (!language && _language) ) {
```
I don't understand the need for the `|| (!language && _language)` part of the condition. If it's to make sure that we print out the WARNING message about setting the language to `nil`, what if you started this function with that check (`if (!self.language) NSLog(@"WARNING...")`) and then you could just make this condition `if ([language isEqualToString:_language] == NO)`?

ws233 replied:
That condition is necessary to eliminate setter run if we try to set nil to the already empty property.
Without this extra condition the condition `[language isEqualToString:_language] == NO` or `[language isEqualToString:_language] == NO` are true if both `language == nil` and `_language == nil`. That's a very rear case, I know. but we have it somewhere in our tests, so I've added it here as well.

kevincon commented on TesseractOCR/G8Tesseract.mm:
```diff
@@ -476,10 +542,19 @@ - (NSUInteger)progress
 - (NSString *)recognizedText
 {
+    if (!self.isEngineConfigured) {
+        NSLog(@"Error! Cannot get recognized text, because engine is not properly configurred!");
```
Typo/grammar, can you rewrite this to read:

> Error! Cannot get recognized text because the Tesseract engine is not properly configured!

kevincon commented on TesseractOCR/G8Tesseract.mm:
```diff
@@ -519,6 +594,11 @@ - (void)analyseLayout
     if (self.layoutAnalysed) return;
+    
+    if (!self.isEngineConfigured) {
+        NSLog(@"Error! cannot perform layout analysis, because the engine is not properly configured!");
```
cannot -> Cannot

Also you can remove the comma in "analysis, because"

kevincon commented on TesseractOCR/G8Tesseract.mm:
```diff
@@ -733,6 +825,10 @@ - (UIImage *)imageWithBlocks:(NSArray *)blocks drawText:(BOOL)drawText threshold
 - (BOOL)recognize
 {
+    if (!self.isEngineConfigured) {
+        return NO;
```
Can you log a warning here that the engine isn't configured properly? This is probably the point at which most users would want to know if the engine is configured or not.

kevincon commented:
Okay I finished reviewing it. I found some typos that should be fixed and I also commented about a few questions that I have, but other than that it looks pretty good. Thanks for working on this. :)

ws233 commented:
@kevincon, I've fixed most of the typos and errors you mentioned and put all of them in the separate commit, so it's easy to review them. When you confirm them, I'll squash them with the corresponding commit.
Also, I've written some comments to the rest of your notes. Pls, read them and let me know what do you think. Thx!

kevincon commented:
Looks good, thanks for fixing those typos. Can you resolve the merge conflicts and do the squash? Then I'll merge it for you.

ws233 commented:
Sure!

kevincon commented:
Thanks! Merging and closing.
